traits.ts

```diff
@@ -0,0 +1,302 @@
+import {
+  Atom,
+  ContractInterfaceFunction,
+  IContractAST,
+  List,
+  TraitReference,
+} from "@hirosystems/clarinet-sdk-wasm";
+import {
+  EnrichedContractInterfaceFunction,
+  ParameterTypeBeforeEnrich,
+  TraitImportType,
+} from "./shared.types";
+import { Simnet } from "@hirosystems/clarinet-sdk";
+
+// TODO: Should handle list, tuple, optional, response nesting.
+export const enrichInterfaceWithTraitData = (
+  ast: IContractAST,
+  traitReferenceMap: Map<string, any>,
+  functionInterfaceList: ContractInterfaceFunction[],
+  targetContractId: string
+): Map<string, EnrichedContractInterfaceFunction[]> => {
+  const enriched = new Map<string, EnrichedContractInterfaceFunction[]>();
+
+  const enrichArgs = (
+    args: any[],
+    functionName: string,
+    paramMap: any,
+    path: string[] = []
+  ): any[] => {
+    return args.map((arg) => {
+      const currentPath = [...path, arg.name];
+      if (arg.type && arg.type.tuple) {
+        return {
+          ...arg,
+          type: {
+            tuple: enrichArgs(
+              arg.type.tuple,
+              functionName,
+              paramMap[arg.name].tuple,
+              currentPath
+            ),
+          },
+        };
+      } else if (paramMap[arg.name]) {
+        const [traitReferenceName, traitReferenceImport] =
+          getTraitReferenceData(ast, functionName, currentPath);
+        if (traitReferenceName && traitReferenceImport) {
+          return {
+            ...arg,
+            type: {
+              trait_reference: {
+                name: traitReferenceName,
+                import: traitReferenceImport,
+              },
+            },
+          };
+        }
+      }
+      return arg;
+    });
+  };
+
+  // Iterate over all functions in the target contract
+  const enrichedFunctions = functionInterfaceList.map((f) => {
+    return {
+      ...f,
+      args: enrichArgs(f.args, f.name, traitReferenceMap.get(f.name)),
+    };
+  });
+
+  enriched.set(targetContractId, enrichedFunctions);
+  return enriched;
+};
+
+export const getTraitReferenceData = (
+  ast: IContractAST,
+  functionName: string,
+  paramPath: string[]
+): [string, any] | [undefined, undefined] => {
+  const findTraitReference = (
+    paramList: any[],
+    path: string[]
+  ): [string, any] | [undefined, undefined] => {
+    for (const param of paramList) {
+      if (!param.expr || !(param.expr as List).List) continue;
```

Comment:
```suggestion
      if (!param.expr || !(param.expr as List).List) { continue; }
```

Would you prefer to stick to curly-braces everywhere? It's easy to mis-indent code and have unexpected side-effects:

```ts
if (condition)
    doSomething();
    doAnotherThing(); // Always executes, regardless of the condition.
```

vs

```ts
if (condition) {
    doSomething();
    doAnotherThing();
}
```